replica/table.cc

```diff
@@ -1519,7 +1519,9 @@ future<std::unordered_map<sstring, table::snapshot_details>> table::get_snapshot
                 auto snapshot_name = de.name;
                 all_snapshots.emplace(snapshot_name, snapshot_details());
                 return lister::scan_dir(snapshots_dir / fs::path(snapshot_name),  { directory_entry_type::regular }, [this, datadir, &all_snapshots, snapshot_name] (fs::path snapshot_dir, directory_entry de) {
-                    return io_check(file_size, (snapshot_dir / de.name).native()).then([this, datadir, &all_snapshots, snapshot_name, snapshot_dir, name = de.name] (auto size) {
+                    return io_check(file_stat, (snapshot_dir / de.name).native(), follow_symlink::no).then([this, datadir, &all_snapshots, snapshot_name, snapshot_dir, name = de.name] (stat_data sd) {
+                        auto size = sd.allocated_size;
```

Comment:
I don't understand how this fixes anything... Isn't the "allocated_size" of a directory also some random potentially-non-zero number?

Also, I think you should move this patch first - to avoid bisection errors in the test that failed before this patch.

Reply:
It's getting the allocated size of regular files only, not of the parent directory.

Reply:
The test didn't fail before this patch, it failed only due to the previous version of this PR.

The issue this fixes is semantic, not an implementation issue.

Reply:
The root problem is that we accumulated the logical size of files in the snapshot directory, but that doesn't reflect the "true" size on disk (that would be returned by the `du` utility, for example)

Reply:
I'm sorry, I don't understand. I think *this* patch fixed some sort of non-useful size counting for an empty directory, and as such why isn't it a problem that it comes after the patch with a test which verifies that an empty directory has size 0?

Reply:
No, the test failure was detected only with the previous version of this series.  The latest version just fixed that but it keeps the original way of returning 0 for the useless case of an empty directory.

Reply:
So what does the last patch (where I wrote this comment) fix?

Reply:
@nyh it fixes the gist of this PR, accounting space used on disk rather than the logical size (a.k.a. length) of the files.

Reply:
I'll commit, because I don't see anything wrong and @kbr- actually already committed a previous version, but I have to say I don't understand at all the "gist of this PR" and I think it might have drifted a little since this PR started:

The commit message says it fixes two issues - #9897 and #9898 - which have little in common. The issue that does discuss "logical" vs "length" of files mentions a possible half-block discrepancy  - is that tiny difference the main point of this PR? But more importantly, the same issue mentions "the size of the directory" as an important thing to account for (I don't know why, isn't it tiny?) but the last version of this patch stopped accounting the size of the directory.

Reply:
> The commit message says it fixes two issues - #9897 and #9898 - which have little in common.

The thing in common is that both caused the `true_snapshots_size` to be wrong.
One reason was effectively multiplying the result by the number of shards,
the other was counting logical size vs. allocated size.

> The issue that does discuss "logical" vs "length" of files mentions a possible half-block discrepancy - is that tiny difference the main point of this PR?

Depending on the file size, for small size the difference between the allocated size vs. the logical size could be relatively huge (e.g. 8K vs. 66 bytes), and once you have hundreds of thousands of those, it accumulates into a meaningful difference.

> But more importantly, the same issue mentions "the size of the directory" as an important thing to account for (I don't know why, isn't it tiny?) but the last version of this patch stopped accounting the size of the directory.

You're right.  I think that for very large directories the storage space for the directory is no longer negligible. Yet, it wasn't counted in the past and unit tests kind of counted on that (in the empty directory case) so we can consider that as part of the unofficial semantics.  I will update the issue about that.

Reply:
> > But more importantly, the same issue mentions "the size of the directory" as an important thing to account for (I don't know why, isn't it tiny?) but the last version of this patch stopped accounting the size of the directory.
> 
> You're right. I think that for very large directories the storage space for the directory is no longer negligible. Yet, it wasn't counted in the past and unit tests kind of counted on that (in the empty directory case) so we can consider that as part of the unofficial semantics. I will update the issue about that.

Unless I'm misunderstanding something, I don't understand how the directory's size can be anything but completely negligible: As I said there is no standard on what the size of a directory is, but it's usually some sort of list of file names. So for a large directory with a lot of files, you'd expect the directory size to be, say, not more than 100 bytes per file. Assuming that the snapshots themselves are **much larger** than 100 bytes each on average, why would the directory size make any difference?

Reply:
I agree it is small relatively to the data size, but it doesn't hurt to be accurate either, especially when the cost of getting this metadata is negligible. This storage space that is consumed on behalf of the snapshot does include the directory holding it, and it will be freed when deleting this snapshot.  If we stored each sstable in a directory of its own (like I suggested in the past with regards to making sstable-move more efficient and naively atomic), I'd still like to account for those directories' storage even if each one could consume only 1-2 blocks in size, like we count the storage consumed by e.g. the TOC file that is also tiny.